File: scripts/yapc2012/erukiti.js
// loop unrolling and thru [2, 3, 5] multiple
// 7, 11, 13, 17, 19, 23, 29, 1

function sieve_fill(primary) {
  if (primary < max_sqrt) {
    var n = primary * 7;
    while (true) {
      if (n >= max) return; sieve[n] = true; n += primary * 4;
      if (n >= max) return; sieve[n] = true; n += primary * 2;
      if (n >= max) return; sieve[n] = true; n += primary * 4;
      if (n >= max) return; sieve[n] = true; n += primary * 2;
      if (n >= max) return; sieve[n] = true; n += primary * 4;
      if (n >= max) return; sieve[n] = true; n += primary * 6;
      if (n >= max) return; sieve[n] = true; n += primary * 2;
      if (n >= max) return; sieve[n] = true; n += primary * 6;
    }
  }
}

function get_primary_sum(nums) {
  if (nums < 5) {
    switch (nums) {
    case 0: return 0;
    case 1: return 2;
    case 2: return 2 + 3;
    case 3: return 2 + 3 + 5;
    case 4: return 2 + 3 + 5 + 7;
    }
  }

  max = Math.floor(nums * Math.log(nums) + nums * Math.log(Math.log(nums))); // Pierre Dusart
  max_sqrt = Math.sqrt(max)
  sieve = new Buffer(max); sieve.fill(0);
  var sum = 2 + 3 + 5;
  nums -= 3;
  var primary = 7;

  while (true) {
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 4;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 2;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 4;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 2;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 4;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 6;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 2;
    if (!sieve[primary]) {sieve_fill(primary); sum += primary; if (--nums == 0) break;} primary += 6;
  }

  return sum;
}

console.log(get_primary_sum(10000));

File: scripts/yapc2012/erukiti.rb
#! /usr/bin/env ruby
#coding: utf-8

include Math

def sieve_fill(sieve, primary, max)
    n = primary * 7
    loop do
      return if n >= max; sieve[n] = true; n += primary * 4
      return if n >= max; sieve[n] = true; n += primary * 2
      return if n >= max; sieve[n] = true; n += primary * 4
      return if n >= max; sieve[n] = true; n += primary * 2
      return if n >= max; sieve[n] = true; n += primary * 4
      return if n >= max; sieve[n] = true; n += primary * 6
      return if n >= max; sieve[n] = true; n += primary * 2
      return if n >= max; sieve[n] = true; n += primary * 6
    end
end

def get_primary_sum(nums)
  return [2, 3, 5, 7][0...nums].inject(:+) if nums < 5

  max = (nums * log(nums) + nums * log(log(nums))).to_i
  max_sqrt = sqrt(max)
  sieve = Array.new(max)
  result = 2 + 3 + 5
  nums -= 3
  primary = 7

  loop do
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 4
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 2
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 4
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 2
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 4
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 6
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 2
    unless sieve[primary] then sieve_fill(sieve, primary, max) if primary < max_sqrt; result += primary; nums -= 1; return result if nums == 0; end; primary += 6
    break if primary >= max_sqrt
  end

  loop do
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 4
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 2
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 4
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 2
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 4
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 6
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 2
    unless sieve[primary] then result += primary; nums -= 1; return result if nums == 0; end; primary += 6
  end
end

#if __FILE__ == $0
  puts get_primary_sum(10000)
#end


